### PR TITLE
MatchFailureException captures an unreified tuple input

### LINKED ISSUE #31482: Should MatchFailureException capture an unreified tuple input?
We have tests like the following:

``` c#
        [Fact]
        public void UnmatchedInput_06()
        {
            var source =
@"using System;
public class C
{
    static void Main()
    {
        Console.WriteLine(M(1, 2));
        try
        {
            Console.WriteLine(M(1, 3));
        }
        catch (MatchFailureException ex)
        {
            Console.WriteLine($""{ex.GetType().Name}({ex.UnmatchedValue})"");
        }
    }
    public static int M(int x, int y) {
        return (x, y) switch { (1, 2) => 3 };
    }
}
namespace System
{
    public class MatchFailureException : InvalidOperationException
    {
        public MatchFailureException() {}
        public MatchFailureException(object unmatchedValue) => UnmatchedValue = unmatchedValue;
        public object UnmatchedValue { get; }
    }
}
";
            var compilation = CreatePatternCompilation(source);
            compilation.VerifyDiagnostics(
                // (17,23): warning CS8509: The switch expression does not handle all possible inputs (it is not exhaustive).
                //         return (x, y) switch { (1, 2) => 3 };
                Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(17, 23)
                );
            CompileAndVerify(compilation, expectedOutput: @"3
MatchFailureException()");
        }
```

Note that the exception throw in this case does not contain the `UnmatchedValue` that was input to the switch expression.  The reason the compiler does not pass that tuple to the exception constructor is that the compiler never constructs the tuple at all.  Rather the compiler produces the decision tree based on the individual values that were the input.  Because of that the generated code does not even have a runtime dependence on the presence of the `ValueTuple` library.

This is consistent with the [design of the `MatchFailureException` type](https://github.com/dotnet/corefx/issues/33284) and the LDM's specification that the captured value shall be provided when it is "trivial" to do so.  However, it would be possible for the compiler to produce code that constructs the tuple just for the failure case.

This is a tracking issue while we seek guidance on how to handle this situation.

### PR DESCRIPTION
**Pull request by gafter**
Fixes #31482

### CODE REVIEW

**Comment by cston** [comment]
>         public SwitchExpressionException() {}

Should this constructor throw an exception, as an extra check?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs:2043 in 5f74d10. [](commit_id = 5f74d1093d1d0c36ad532ef6df9028f44e56aace, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs [hunk 1]
```diff
@@ -2052,7 +2052,7 @@ public class SwitchExpressionException : InvalidOperationException
                 //             _ = (1, 2) switch { (3, 4) => 1 };
                 Diagnostic(ErrorCode.WRN_SwitchExpressionNotExhaustive, "switch").WithLocation(8, 24)
                 );
-            CompileAndVerify(compilation, expectedOutput: "SwitchExpressionException()");
+            CompileAndVerify(compilation, expectedOutput: "SwitchExpressionException((1, 2))");
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Please add a test with missing `ValueTuple` type.
Maybe testing a long tuple could be good too.

Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Also, the meeting notes from [November](https://github.com/dotnet/csharplang/blob/4811c98606902cc85822e24c5dc5565a2bdadcab/meetings/2018/LDM-2018-11-14.md) list the exception as being called `MatchFailureException`. Did I miss a rename?

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs]: Yes; see [https://github.com/dotnet/roslyn/issues/31911](https://github.com/dotnet/roslyn/issues/31911)

---
In reply to: [249960714](https://github.com/dotnet/roslyn/pull/32588#discussion_r249960714) [](ancestors = 249960714)

**Comment by gafter** [comment]
> did you already block switch expressions in expression trees?

Yes.  See https://github.com/dotnet/roslyn/blob/5f3e87af8dfb356a16b1d9caf36cea129545be21/src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_ExpressionTrees.cs#L167-L175

**Pull request merged**